src/main/java/com/google/devtools/build/lib/packages/BuildType.java

```diff
@@ -558,6 +558,23 @@ public void repr(SkylarkPrinter printer) {
       this(x, what, context, originalType, "");
     }
 
+    /**
+     * This class provides more precise error messaging for type failures on select()s by
+     * highlighting which select() branch has the failure.
+     */
+    private static class CustomToString {
```

Comment:
Hope I'm not too late, but I'm not sure this is accurate - my understanding is that the error is actually a type-checking error, unrelated to actual branch selection. The message suggests that a branch was actually selected, which is incorrect, I believe. (Test this by putting 99 in any select expression that wants some other type).  In that case I think the message should indicate more clearly that it's a typing problem, maybe something like "Ill-typed select expression: branch %s value has type %s, expected type is %s"

Reply:
Yes, it's a type checking error. As it stands now the error would look like:

`//testapp:gen: expected value of type 'string' for attribute 'cmd' in 'genrule' rule over select() branch '//conditions:default', but got ["echo hi > $@"] (list)`

Happy to try to clarify that further, but I'd rather do this within the existing grammar of

`//testapp:gen: expected value of type 'string' <INSERT WHATEVER MESSAGE WE WANT HERE>, but got ["echo hi > $@"] (list)`

This is because the full error message is generated [here](https://github.com/bazelbuild/bazel/blob/1d273cd492f79f018ee05bf6019c6f9caba34673/src/main/java/com/google/devtools/build/lib/syntax/Type.java#L239) (`what` is the variable we can freely mess with). This applies to *all* type checking, whether `select`s are there or not. It seems too invasive to change this API just for this case.

Any ideas that fit within the grammar?